File: NET_4.5.2/C1.WPF.FlexGrid/VB/FilterRow/FilterRowCellFactory.vb
Imports System.Text
Imports System.Text.RegularExpressions
Imports System.ComponentModel
Imports C1.WPF.FlexGrid

Namespace FilterRow
	Friend Class FilterRowCellFactory
		Inherits CellFactory
		' ** fields
		Private Shared _editorPadding As New Thickness(0, 0, 2, 0)

		' ** properties

		''' <summary>
		''' Set this property to true to handle the '%' character as a wildcard
		''' when filtering.
		''' For example, 
		'''     StartsWith%
		'''     %Endswith
		'''     %Contains%
		''' </summary>
		Public Property UseRegEx() As Boolean

		' ** overrides
		Public Overrides Sub CreateColumnHeaderContent(ByVal grid As C1FlexGrid, ByVal bdr As Border, ByVal range As CellRange)
			' create the filter row if it isn't there
			Dim rows = grid.ColumnHeaders.Rows
			If rows.Count = 0 OrElse Not(TypeOf rows(rows.Count - 1) Is GridFilterRow) Then
				rows.Add(New GridFilterRow())
			End If

			' add filter editors to filter row
			Dim fr = TryCast(rows(range.Row), GridFilterRow)
			If fr IsNot Nothing Then
				bdr.Child = New FilterEditor(fr, grid.Columns(range.Column))
				bdr.Padding = _editorPadding
			Else
				MyBase.CreateColumnHeaderContent(grid, bdr, range)
			End If
		End Sub
		Public Overrides Sub CreateTopLeftContent(ByVal grid As C1FlexGrid, ByVal bdr As Border, ByVal range As CellRange)
			Dim rows = grid.ColumnHeaders.Rows
			Dim fr = TryCast(rows(range.Row), GridFilterRow)
			If fr IsNot Nothing AndAlso range.Column = grid.TopLeftCells.Columns.Count - 1 Then
				bdr.Child = New ClearFilterButton(fr)
			Else
				MyBase.CreateTopLeftContent(grid, bdr, range)
			End If
		End Sub
		Public Overrides Function IsSortSymbolRow(ByVal grid As C1FlexGrid, ByVal range As CellRange) As Boolean
			' show sort symbols on the column header row above the filter row
			Return range.TopRow = grid.ColumnHeaders.Rows.Count - 2
		End Function
	End Class
	Public Class GridFilterRow
		Inherits Row
		' ** fields
		Private _filterArgs As New Dictionary(Of Column, String)()
		Private _rxPattern As New Dictionary(Of Column, String)()
		Private _useRegex As Boolean

		' ** object model
		Public Function GetFilterArgument(ByVal col As Column) As String
			Dim text As String
			Return If(_filterArgs.TryGetValue(col, text), text, String.Empty)
		End Function
		Public Sub SetFilterArgument(ByVal col As Column, ByVal arg As String)
			If arg <> GetFilterArgument(col) Then
				' save argument
				_filterArgs(col) = arg

				' save regex pattern to support wildcards
				arg = If(arg IsNot Nothing, arg.Replace("%", ".*"), arg)
				_rxPattern(col) = String.Format("^{0}$", arg)

				' apply the filter
				ApplyFilter()
			End If
		End Sub

		' ** implementation
		Private Sub ApplyFilter()
			' check whether to use wildcards when matching
			Dim cf = TryCast(Grid.CellFactory, FilterRowCellFactory)
			_useRegex = cf IsNot Nothing AndAlso cf.UseRegEx

			' apply filter to grid's ICollectionView
			Dim view = TryCast(Grid.ItemsSource, ICollectionView)
			If view Is Nothing OrElse (Not view.CanFilter) Then
				' apply the filter in unbound mode
				For Each row In Grid.Rows
					row.Visible = FilterPredicate(row)
				Next row

				' done
				Return
			End If

			' can't apply filters while editing or adding items
			Dim edt = TryCast(view, IEditableCollectionView)
			If edt IsNot Nothing Then
				If edt.IsEditingItem Then
					edt.CommitEdit()
				End If

				If edt.IsAddingNew Then
					edt.CommitNew()
				End If
			End If

			' apply the filter in bound mode
			Try
				view.Filter = AddressOf FilterPredicate
			Catch x As Exception
				MessageBox.Show("Failed to apply the filter:" & vbCrLf, x.Message)
			End Try
		End Sub
		Private Function FilterPredicate(ByVal item As Object) As Boolean
			' get item type to retrieve properties
			Dim itemType = item.GetType()

			' in unbound mode, the item is the row itself
			Dim unbound As Boolean = TypeOf item Is Row

			' check filter values applied to each column
			For Each kv In _filterArgs
				' no filter condition on this column? continue
				If String.IsNullOrEmpty(kv.Value) Then
					Continue For
				End If

				' get cell value
				Dim stringValue = String.Empty
				Dim row = TryCast(item, Row)
				If row IsNot Nothing Then
					' handle unbound mode
					Dim value = row(kv.Key)
					stringValue = If(value Is Nothing, String.Empty, value.ToString())
				Else
					' get property name
					Dim propName = kv.Key.BoundPropertyName
					If String.IsNullOrEmpty(propName) Then
						Continue For
					End If

					' get property info
					Dim prop = itemType.GetProperty(propName)
					If prop Is Nothing Then
						Continue For
					End If

					' get property value
					Dim value = prop.GetValue(item, Nothing)
					stringValue = If(value Is Nothing, String.Empty, value.ToString())
				End If

				If _useRegex Then
					' test using regular expressions (support wildcards)
					If Not Regex.IsMatch(stringValue, _rxPattern(kv.Key), RegexOptions.IgnoreCase) Then
						Return False
					End If
				Else
					' text using simple 'contains' 
					If stringValue.IndexOf(kv.Value, StringComparison.OrdinalIgnoreCase) < 0 Then
						Return False
					End If
				End If
			Next kv

			' item passed all filter conditions
			Return True
		End Function
	End Class
End Namespace

File: NET_4.6.2/C1.WPF.FlexGrid/CS/FilterRow/FilterRow/FilterRowCellFactory.cs
using System;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Media;
using System.Windows.Media.Imaging;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using System.ComponentModel;
using C1.WPF.FlexGrid;

namespace FilterRow
{
    class FilterRowCellFactory : CellFactory
    {
        // ** fields
        static Thickness _editorPadding = new Thickness(0, 0, 2, 0);

        // ** properties

        /// <summary>
        /// Set this property to true to handle the '%' character as a wildcard
        /// when filtering.
        /// For example, 
        ///     StartsWith%
        ///     %Endswith
        ///     %Contains%
        /// </summary>
        public bool UseRegEx { get; set; }

        // ** overrides
        public override void CreateColumnHeaderContent(C1FlexGrid grid, System.Windows.Controls.Border bdr, CellRange range)
        {
            // create the filter row if it isn't there
            var rows = grid.ColumnHeaders.Rows;
            if (rows.Count == 0 || !(rows[rows.Count - 1] is GridFilterRow))
            {
                rows.Add(new GridFilterRow());
            }

            // add filter editors to filter row
            var fr = rows[range.Row] as GridFilterRow;
            if (fr != null)
            {
                bdr.Child = new FilterEditor(fr, grid.Columns[range.Column]);
                bdr.Padding = _editorPadding;
            }
            else
            {
                base.CreateColumnHeaderContent(grid, bdr, range);
            }
        }
        public override void CreateTopLeftContent(C1FlexGrid grid, System.Windows.Controls.Border bdr, CellRange range)
        {
            var rows = grid.ColumnHeaders.Rows;
            var fr = rows[range.Row] as GridFilterRow;
            if (fr != null && range.Column == grid.TopLeftCells.Columns.Count - 1)
            {
                bdr.Child = new ClearFilterButton(fr);
            }
            else
            {
                base.CreateTopLeftContent(grid, bdr, range);
            }
        }
        public override bool IsSortSymbolRow(C1FlexGrid grid, CellRange range)
        {
            // show sort symbols on the column header row above the filter row
            return range.TopRow == grid.ColumnHeaders.Rows.Count - 2;
        }

        public void RefreshFilter(C1FlexGrid grid)
        {
            var rows = grid.ColumnHeaders.Rows;
            var filterRow = rows[rows.Count-1] as GridFilterRow;

            filterRow?.ApplyFilter();
        }
    }
    public class GridFilterRow : Row
    {
        // ** fields
        Dictionary<Column, string> _filterArgs = new Dictionary<Column, string>();
        Dictionary<Column, string> _rxPattern = new Dictionary<Column, string>();
        bool _useRegex;

        // ** object model
        public string GetFilterArgument(Column col)
        {
            string text;
            return _filterArgs.TryGetValue(col, out text)
                ? text
                : string.Empty;
        }
        public void SetFilterArgument(Column col, string arg)
        {
            if (arg != GetFilterArgument(col))
            {
                // save argument
                _filterArgs[col] = arg;

                // save regex pattern to support wildcards
                arg = arg != null ? arg.Replace("%", ".*") : arg;
                _rxPattern[col] = string.Format("^{0}$", arg);

                // apply the filter
                ApplyFilter();
            }
        }

        // ** implementation
        internal void ApplyFilter()
        {
            // check whether to use wildcards when matching
            var cf = Grid.CellFactory as FilterRowCellFactory;
            _useRegex = cf != null && cf.UseRegEx;

            // apply filter to grid's ICollectionView
            var view = Grid.ItemsSource as ICollectionView;
            if (view == null || !view.CanFilter)
            {
                // apply the filter in unbound mode
                foreach (var row in Grid.Rows)
                {
                    row.Visible = FilterPredicate(row);
                }

                // done
                return;
            }

            // can't apply filters while editing or adding items
            var edt = view as IEditableCollectionView;
            if (edt != null)
            {
                if (edt.IsEditingItem)
                    edt.CommitEdit();

                if (edt.IsAddingNew)
                    edt.CommitNew();
            }

            // apply the filter in bound mode
            try
            {
                view.Filter = FilterPredicate;
            }
            catch (Exception x)
            {
                MessageBox.Show("Failed to apply the filter:\r\n", x.Message);
            }
        }
        bool FilterPredicate(object item)
        {
            // get item type to retrieve properties
            var itemType = item.GetType();

            // in unbound mode, the item is the row itself
            bool unbound = item is Row;

            // check filter values applied to each column
            foreach (var kv in _filterArgs)
            {
                // no filter condition on this column? continue
                if (string.IsNullOrEmpty(kv.Value))
                {
                    continue;
                }

                // get cell value
                var stringValue = string.Empty;
                var row = item as Row;
                if (row != null)
                {
                    // handle unbound mode
                    var value = row[kv.Key];
                    stringValue = value == null ? string.Empty : value.ToString();
                }
                else
                {
                    // get property name
                    var propName = kv.Key.BoundPropertyName;
                    if (string.IsNullOrEmpty(propName))
                    {
                        continue;
                    }

                    // get property info
                    var prop = itemType.GetProperty(propName);
                    if (prop == null)
                    {
                        continue;
                    }

                    // get property value
                    var value = prop.GetValue(item, null);
                    stringValue = value == null ? string.Empty : value.ToString();
                }

                if (_useRegex)
                {
                    // test using regular expressions (support wildcards)
                    if (!Regex.IsMatch(stringValue, _rxPattern[kv.Key], RegexOptions.IgnoreCase))
                    {
                        return false;
                    }
                }
                else
                {
                    // text using simple 'contains' 
                    if (stringValue.IndexOf(kv.Value, StringComparison.OrdinalIgnoreCase) < 0)
                    {
                        return false;
                    }
                }
            }

            // item passed all filter conditions
            return true;
        }
    }
}
